lib/datadog/statsd/message_buffer.rb

```diff
@@ -28,30 +35,53 @@ def add(message)
         return nil unless message_size > 0 # to avoid adding empty messages to the buffer
         return nil unless ensure_sendable!(message_size)
 
-        flush if should_flush?(message_size)
+        @mon.synchronize {
+          raise Error, 'buffer is closed' if @closed
+          flush if should_flush?(message_size)
 
-        buffer << "\n" unless buffer.empty?
-        buffer << message
+          buffer << "\n" unless buffer.empty?
+          buffer << message
 
-        @message_count += 1
+          @message_count += 1
 
-        # flush when we're pretty sure that we won't be able
-        # to add another message to the buffer
-        flush if preemptive_flush?
+          # flush when we're pretty sure that we won't be able
+          # to add another message to the buffer
+          flush if preemptive_flush?
+        }
 
         true
       end
 
       def reset
-        clear_buffer
-        connection.reset_telemetry
+        @mon.synchronize {
+          clear_buffer
+          connection.reset_telemetry
+          @flush_thread = create_flush_thread if @flush_interval
```

Comment:
This changes the semantics of `reset` such that it can only be called when the buffer is in a closed state (has no running flush thread).  As currently used in `Sender`, that's not quite true: it's called when the buffer is not closed, but coincidentally in a condition where it's likely that the flush thread has been killed by a `fork`.

Maybe let's do a similar thing, where the flush thread is only created if it's `nil` or if it's not alive, similar to
https://github.com/DataDog/dogstatsd-ruby/blob/48f99e869e969431db3264dd50f1bcd7125de919/lib/datadog/statsd/sender.rb#L64-L64

Reply:
I think you suggest I write as follows, but I thought the change like https://github.com/DataDog/dogstatsd-ruby/pull/154/commits/89811077faf05e0f16676a0e4f94dcff338019ac would be better:

```ruby
if @flush_interval && (@flush_thread.nil? || @flush_thread.stop?)
  @flush_thread = create_flush_thread
end
```

You can refer to the commit message for more details. How about this?

Reply:
That makes great sense - thank you!